src/aws_encryption_sdk/materials_managers/__init__.py

```diff
@@ -40,38 +48,250 @@ class EncryptionMaterialsRequest(object):
+
+        for flag in required_flags:
+            if flag not in keyring_trace.flags:
+                raise InvalidKeyringTraceError("Keyring flags do not match action.")
+
+        if keyring_trace.wrapping_key != data_encryption_key.key_provider:
+            raise InvalidKeyringTraceError("Keyring trace does not match data key provider.")
+
+        if len(data_encryption_key.data_key) != self.algorithm.kdf_input_len:
+            raise InvalidDataKeyError(
+                "Invalid data key length {actual} must be {expected}.".format(
+                    actual=len(data_encryption_key.data_key), expected=self.algorithm.kdf_input_len
+                )
+            )
+
+    def _add_data_encryption_key(self, data_encryption_key, keyring_trace, required_flags):
+        # type: (Union[DataKey, RawDataKey], KeyringTrace, Iterable[KeyringTraceFlag]) -> None
+        """Add a plaintext data encryption key.
+
+        :param RawDataKey data_encryption_key: Data encryption key
+        :param KeyringTrace keyring_trace: Trace of actions that a keyring performed
+          while getting this data encryption key
+        :raises AttributeError: if data encryption key is already set
+        :raises InvalidKeyringTraceError: if keyring trace does not match required actions
+        :raises InvalidKeyringTraceError: if keyring trace does not match data key provider
+        :raises InvalidDataKeyError: if data key length does not match algorithm suite
+        """
+        self._validate_data_encryption_key(
+            data_encryption_key=data_encryption_key, keyring_trace=keyring_trace, required_flags=required_flags
+        )
+
+        data_key = _data_key_to_raw_data_key(data_key=data_encryption_key)
+
+        super(CryptographicMaterials, self).__setattr__("data_encryption_key", data_key)
+        self._keyring_trace.append(keyring_trace)
+
+    @property
+    def keyring_trace(self):
+        """Return a read-only version of the keyring trace.
+
+        :rtype: tuple
+        """
+        return tuple(self._keyring_trace)
+
+
+@attr.s(hash=False, init=False)
+class EncryptionMaterials(CryptographicMaterials):
     """Encryption materials returned by a crypto material manager's `get_encryption_materials` method.
 
     .. versionadded:: 1.3.0
 
-    :param algorithm: Algorithm to use for encrypting message
-    :type algorithm: aws_encryption_sdk.identifiers.Algorithm
-    :param data_encryption_key: Plaintext data key to use for encrypting message
-    :type data_encryption_key: aws_encryption_sdk.structures.DataKey
-    :param encrypted_data_keys: List of encrypted data keys
-    :type encrypted_data_keys: list of `aws_encryption_sdk.structures.EncryptedDataKey`
+    .. versionadded:: 1.5.0
+
+        The **keyring_trace** parameter.
+
+    .. versionadded:: 1.5.0
+
+        Most parameters are now optional.
+
+    :param Algorithm algorithm: Algorithm to use for encrypting message
```

Comment:
tiny nit, are we not moving to "Algorithm Suite"?

Reply:
This was to maintain backwards compatibility with the old `EncryptionMaterials` API, but you're right. I think adding something like the `DecryptionMaterials.data_key` redirect is the right thing to do here.

Reply:
Actually, looking back through the API, I thought we had moved to `algorithm_suite` elsewhere but we have not; it looks like the only other places we have moved to that name is in internal APIs.

For consistency within this client, I'm going to leave this as-is.